tls/s2n_security_policies.c

```diff
@@ -1465,16 +1471,40 @@ S2N_RESULT s2n_security_policy_validate_cert_signature(
 {
     RESULT_ENSURE_REF(info);
     RESULT_ENSURE_REF(security_policy);
-    const struct s2n_signature_preferences *sig_preferences =
-            security_policy->certificate_signature_preferences;
-
-    RESULT_ENSURE_REF(sig_preferences);
+    const struct s2n_signature_preferences *sig_preferences = security_policy->certificate_signature_preferences;
 
-    for (size_t i = 0; i < sig_preferences->count; i++) {
-        if (sig_preferences->signature_schemes[i]->libcrypto_nid == info->signature_nid) {
-            return S2N_RESULT_OK;
+    if (sig_preferences != NULL) {
+        for (size_t i = 0; i < sig_preferences->count; i++) {
+            if (sig_preferences->signature_schemes[i]->libcrypto_nid == info->signature_nid) {
+                return S2N_RESULT_OK;
+            }
         }
+
+        RESULT_BAIL(S2N_ERR_SECURITY_POLICY_INCOMPATIBLE_CERT);
     }
 
-    RESULT_BAIL(S2N_ERR_CERT_UNTRUSTED);
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_security_policy_validate_certificate_chain(
+        const struct s2n_security_policy *security_policy,
+        const struct s2n_cert_chain_and_key *cert_key_pair)
+{
+    RESULT_ENSURE_REF(security_policy);
+    RESULT_ENSURE_REF(cert_key_pair);
+    RESULT_ENSURE_REF(cert_key_pair->cert_chain);
+
+    if (!security_policy->certificate_preferences_apply_locally) {
+        return S2N_RESULT_OK;
+    }
+
+    struct s2n_cert *current = cert_key_pair->cert_chain->head;
+    while (current != NULL) {
+        if (!current->info.self_signed) {
```

Comment:
If you're still waffling, I'd lean towards validating the self-signed one. Madeleine makes a good point about simplicity, and it's easier to relax a restriction later than to tighten it.